File: src/Warehouse.Core/Markets/Concrete/Okx/DependencyInjection.cs
using System.Net;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Polly;
using Warehouse.Core.Infrastructure.WebSockets;
using Warehouse.Core.Markets.Concrete.Okx.Services;
using Warehouse.Core.Markets.Contracts;
using Warehouse.Core.Markets.Domain;

namespace Warehouse.Core.Markets.Concrete.Okx;

public static class DependencyInjection
{
    public static IServiceCollection AddOkxSupport(this IServiceCollection services, IConfiguration configuration)
    {
        services.Configure<MarketCredentials>(configuration.GetSection(nameof(MarketCredentials)));
        services.AddSingleton<IWebSocketClient, WebSocketClient>();
        services.AddSingleton<OkxHeartbeatService>();
        services.AddScoped<OkxHttpService>();
        services.AddScoped<IMarketBalanceProvider, OkxBalanceProvider>();
        services.AddSingleton<OkxMarketAdapter>();
        services.AddHostedService<OkxSynchronizationWorker>();

        services.AddHttpClient(
                "Okx",
                client =>
                {
                    client.BaseAddress = new Uri("https://www.okx.com/");
                    client.Timeout = TimeSpan.FromSeconds(30);
                    client.DefaultRequestHeaders.Add("User-Agent", "Analyzer/1.0");
                })
            .ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler
            {
                AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate
            })
            .AddPolicyHandler((provider, _) =>
            {
                ILogger<HttpClient> logger = provider.GetRequiredService<ILogger<HttpClient>>();
                return Policy.HandleResult<HttpResponseMessage>(r => !r.IsSuccessStatusCode)
                    .WaitAndRetryAsync(
                        3,
                        retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
                        (_, timespan, retryCount, _) =>
                        {
                            logger.LogWarning("Retry attempt {RetryCount} after {TimeSpan} seconds", retryCount, timespan);
                        });
            })
            .AddPolicyHandler(_ => Policy.TimeoutAsync<HttpResponseMessage>(10));

        return services;
    }
}

File: src/Warehouse.Core/Markets/Concrete/Okx/DependencyInjection.fs
namespace Warehouse.Core.Markets.Concrete.Okx

open System
open System.Net
open System.Net.Http
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Logging
open Warehouse.Core.Functional.Infrastructure.WebSockets
open Warehouse.Core.Functional.Markets.Concrete.Okx.Services
open Warehouse.Core.Functional.Markets.Domain
open Warehouse.Core.Functional.Orders.Contracts
open Warehouse.Core.Markets.Concrete.Okx.Services
open Polly

module DependencyInjection =
    let AddOkxSupport (services: IServiceCollection, configuration: IConfiguration) =
        services.Configure<MarketCredentials>(configuration.GetSection(nameof MarketCredentials)) |> ignore
        services.AddSingleton<IWebSocketClient, WebSocketClient>() |> ignore
        services.AddSingleton<OkxHeartbeatService>() |> ignore
        services.AddScoped<OkxHttpService>() |> ignore
        services.AddScoped<IMarketOrderProvider, OkxMarketOrderProvider>() |> ignore
        services.AddSingleton<OkxMarketAdapter>() |> ignore
        services.AddHostedService<OkxSynchronizationWorker>() |> ignore

        services
            .AddHttpClient(
                "Okx",
                fun (client: HttpClient) ->
                    client.BaseAddress <- Uri("https://www.okx.com/")
                    client.Timeout <- TimeSpan.FromSeconds(30.0)
                    client.DefaultRequestHeaders.Add("User-Agent", "Analyzer/1.0")
            )
            .ConfigurePrimaryHttpMessageHandler(
                Func<HttpMessageHandler>(fun () ->
                    new HttpClientHandler(
                        AutomaticDecompression = (DecompressionMethods.GZip ||| DecompressionMethods.Deflate)
                    ))
            )
            .AddPolicyHandler(fun (provider: IServiceProvider) (_: HttpRequestMessage) ->
                let logger = provider.GetRequiredService<ILogger<HttpClient>>()

                Policy
                    .HandleResult<HttpResponseMessage>(fun (r: HttpResponseMessage) -> not r.IsSuccessStatusCode)
                    .WaitAndRetryAsync(
                        3,
                        (fun retryAttempt -> TimeSpan.FromSeconds(Math.Pow(2.0, float retryAttempt))),
                        (fun result timespan retryCount ctx ->
                            logger.LogWarning(
                                "Retry attempt {RetryCount} after {TimeSpan} seconds",
                                retryCount,
                                timespan
                            ))
                    )
                :> IAsyncPolicy<HttpResponseMessage>)
            .AddPolicyHandler(fun _ ->
                Policy.TimeoutAsync<HttpResponseMessage>(10) :> IAsyncPolicy<HttpResponseMessage>)
        |> ignore

        services
